### PR TITLE
feat: Gradual rollout of new features

### PR DESCRIPTION
**Pull request by Jonasdoubleyou**
Allow features to be gradually enabled for a random subset of users, which can gradually be increased.

Participants are tracked via Datadog RUM, to be able to identify issues with new features.

resolves https://github.com/corona-school/project-user/issues/1238

### CODE REVIEW

**Review by JeangelLF — APPROVED** [review]
Very interesting, thank you!

File: src/config.ts [hunk 1]
```diff
@@ -1,3 +1,30 @@
...
+    let randomString = localStorage.getItem(key);
+    if (!randomString) {
+        randomString = Math.floor(Math.random() * 100).toString();
+        localStorage.setItem(key, randomString);
+    }
+
+    let random = parseInt(randomString, 10);
+    let rolloutRatio = parseInt(rolloutRatioString ?? '0', 10);
+
+    const chosen = random < rolloutRatio;
```
Comment (dhenkel92) [inline src/config.ts:22]: This will be false if random is 100, should be `random <= rolloutRatio`

image

**Review by dhenkel92 — COMMENTED** [review]
Pretty cool 🙂 👍

File: src/config.ts (hunk 1 above)
Comment (Jonasdoubleyou) [reply src/config.ts:22]: Argh, floating point math. Math.random() is [0, 1), so *100 should be [0, 100) but I guess it is actually [0, 100 + Epsilon) and 100 + Epsilon might not be < 100 ...

Comment (Jonasdoubleyou) [reply src/config.ts:22]: On the contrary, if the rollout ratio is 0, <= would also be wrong ... I guess we should just turn the rollout ratio to 101, or just remove the toggle if we turn it on for all users anyways

Comment (dhenkel92) [reply src/config.ts:22]: That's true. What about being more explicit in code? We could set chose to true if rolloutRation is 100% and to false if it's zero. It's a couple more lines of code, but ensures that these boundaries are always given, no matter what `random` is.

**Pull request merged**